kubeflow/application/application.libsonnet

```diff
@@ -246,15 +159,201 @@
           params.components,
     }.return,
 
+    local groupByName(resources) = {
+      [resource.name]+: resource
+      for resource in resources
+    },
+
+    local groupByResource(resources) = {
+      local getKey(resource) = {
+        return::
+          //resource.kind + "." + resource.apiVersion,
+          resource.kind,
+      }.return,
+      local getValue(resource) = {
+        return::
+          { [resource.metadata.name]+: resource },
+      }.return,
+      return:: util.foldl(getKey, getValue, resources),
+    }.return,
+
+    local clusterScope(resource) = {
+      return::
+        if (std.objectHas(resource, "metadata") &&
+            !std.objectHas(resource.metadata, "namespace")) then
+          true
+        else
+          false,
+    }.return,
+    local namespacedScope(resource) = {
+      return:: clusterScope(resource) == false,
+    }.return,
     local tuples = std.flattenArrays(std.map(perComponent, getComponents)),
-    local components = std.map(byResource, tuples),
+    local resources = std.map(byResource, tuples),
+    local namespacedResources = std.filter(namespacedScope, resources),
+    local clusterResources = std.filter(clusterScope, resources),
+    local groupedNamespacedResources = groupByResource(namespacedResources),
+
+    local syncApplicationTemplate = importstr "sync-application.template",
+    local syncApplication = syncApplicationTemplate % {
+      resources: std.manifestJsonEx(namespacedResources, "  "),
+      groupedResources: std.manifestJsonEx(groupedNamespacedResources, "  "),
+    },
+
+    local applicationConfigMap = {
+      apiVersion: "v1",
+      kind: "ConfigMap",
+      metadata: {
+        name: "application-controller-hooks",
+        namespace: params.namespace,
+      },
+      data: {
+        "sync-application.jsonnet": syncApplication,
+        "util.libsonnet": importstr "kubeflow/common/util.libsonnet",
+      },
+    },
+    applicationConfigMap:: applicationConfigMap,
+
+    local applicationDeployment = {
```

Comment:
So the application-controller is metacontroller? I thought we had a separate package now for deploying metacontroller?

Reply:
It relies on metacontroller but the application-controller deploys a Deployment, Service and ConfigMap - this is normal for any metacontroller instance (profiles, jupyter, ...).

So metacontroller component deploys 
- CompositeController CRD
- DecoratorController CRD
- ServiceAccount
- ClusterRole
- ClusterRoleBinding
- Statefule